File: php/day10_part2_2022.php
<?php
function readAll($path) {
    return file_get_contents($path);
}

function myAbs($x) {
    return $x < 0 ? -$x : $x;
}

$x = [1];
foreach (explode("\n", readAll("input.txt")) as $line) {
    switch ($line) {
        case "noop":
            $x[] = $x[count($x) - 1];
            break;
        default:
            $n = intval(substr($line, 5));
            $x[] = $x[count($x) - 1];
            $x[] = $x[count($x) - 1] + $n;
            break;
    }
}

$grid = [];
for ($i = 0; $i < count($x); $i++) {
    $crtx = $i % 40;
    $crty = (int)($i / 40);
    if (myAbs($crtx - $x[$i]) <= 1) {
        $grid[$crtx . "," . $crty] = "#";
    } else {
        unset($grid[$crtx . "," . $crty]);
    }
}

for ($y = 0; $y < 6; $y++) {
    for ($x = 0; $x < 40; $x++) {
        echo isset($grid[$x . "," . $y]) ? "#" : ".";
    }
    echo "\n";
}
?>

File: cpp/day10_part2_2022.cpp

#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <cmath>
#include <array>
#include <algorithm>

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    std::vector<int> x;
    x.push_back(1);

    std::ifstream inputFile("input.txt");
    std::string line;

    while (std::getline(inputFile, line)) {
        if (line == "noop") {
            x.push_back(x.back());
        } else {
            int n = std::stoi(line.substr(line.find(' ') + 1));
            x.push_back(x.back());
            x.push_back(x.back() + n);
        }
    }
    inputFile.close();

    std::array<std::array<char, 40>, 6> grid;

    for (int r = 0; r < 6; ++r) {
        for (int c = 0; c < 40; ++c) {
            grid[r][c] = '.';
        }
    }

    int max_cycles_for_screen = 6 * 40;
    for (int i = 0; i < std::min((int)x.size(), max_cycles_for_screen); ++i) {
        int crtx = i % 40;
        int crty = i / 40;
        int val = x[i];

        if (std::abs(crtx - val) <= 1) {
            grid[crty][crtx] = '#';
        }
    }

    for (int r = 0; r < 6; ++r) {
        for (int c = 0; c < 40; ++c) {
            std::cout << grid[r][c];
        }
        std::cout << '\n';
    }

    return 0;
}

